libs/proxy/tokio-postgres2/src/connection.rs

```diff
@@ -289,18 +292,30 @@ where
         &mut self,
         cx: &mut Context<'_>,
     ) -> Poll<Option<Result<AsyncMessage, Error>>> {
-        let message = self.poll_read(cx)?;
-        let want_flush = self.poll_write(cx)?;
-        if want_flush {
-            self.poll_flush(cx)?;
+        if self.state != State::Closing {
+            // if the state is still active, try read from and write to postgres.
+            let message = self.poll_read(cx)?;
+            let closing = self.poll_write(cx)?;
+            if let Poll::Ready(WriteReady::Terminating) = closing {
```

Comment:
We're switching from `poll_read` to `poll_write` to signal when a connection is closed and then stop reading remaining messages. Is that safe?

Reply:
`poll_write` was always the function that set `self.state = State::Closing`. `poll_read` never updated `self.state`.

Reply:
Right, looks good I guess. I still think that the old code keeps reading remaining messages until the framed `Stream` returns `None` and we get an error of `Kind::Closed`. That's probably not right either.

Reply:
I tried to make each commit very small, where each commit should be easy to verify it doesn't change the behaviour.

If you think one of the commits does change the behaviour, then I can try and fix it. I have avoided trying to actually fix any problems here (although I have not found any problems so far)